docs/adr/adr-009-telemetry.md

```diff
@@ -0,0 +1,226 @@
+  * The richest UI
+  * Tight to Clickhouse DB
+  * Made by OpenTelemetry
+  * The most lightweight
+* [Jaeger](https://www.jaegertracing.io/)
+  * The most mature
+  * Started by Uber, now supported by CNCF
+  * Supports multiple storages(ScyllaDB, InfluxDB, Amazon DynamoDB)
+* [Graphana Tempo](https://grafana.com/oss/tempo/)
+  * Deep integration with Graphana/Prometheus
+  * Relatively new (~2 years)
+  * Uses Azure, GCS, S3 or local disk for storage
+
+Each of these backends can be used independently and depending on the use case. For us, these are main use cases: 
+* Local development/debugging for the private network or even public network setup
+> I am personally planning to set up the lightweight Uptrace for the local light node. Just to play around and observe 
+> things
+* Data collection from the Testground test runs
+* Bootstrappers monitoring infrastructure
+* Data collection from the incentivized testnet participants
+
+There is no strict decision on which of these backends and where to use. People taking ownership of any listed vectors
+are free to use any recommended solution or any unlisted.
+
+#### Metrics Backend
+
+// WIP
+
+## Design
+
+### Tracing
+
+Tracing allows to see _how_ any process progresses through different modules, APIs and networks, as well as timings of 
+each operation and any events or errors as they occur.
+
+A visual example of a generic tracing dashboard provided via [Uptrace](https://uptrace.dev/) backend
+![tracing](img/tracing-dashboard.png)
+
+Mainly, for "ShrEx" and reconstruction analysis we need to know if the reconstruction succeeded and the time it took.
+The tracing in this case would provide all three metrics for the whole reconstruction operation and for each sub operation of each
+process.
+
+#### Spans
+Span represents an operation (unit of work) in a trace. They keep the time when operation _started_ and _ended_. Any 
+additional user defined _attributes_, operation status(success or error with an error itself) and events/logs that 
+may happen during the operation.
+
+Spans also form a parent tree, meaning that each span associated to a process can have multiple sub processes or child
+spans and vise-versa. Altogether, this feature allows to see the whole trace of execution of any part of the system, no 
+matter how complex it is. This is exactly what we need to analyze our reconstruction performance.
+
+An example of a root span in `ipld.Retriever`:
+```go
+import "go.opentelemetry.io/otel"
+
+var tracer = otel.Tracer("share")
+
+func (r *Retriever) Retrieve(ctx context.Context, dah *da.DataAvailabilityHeader) (*rsmt2d.ExtendedDataSquare, error) {
+  ctx, span := tracer.Start(ctx, "retrieve-square")
+  defer span.End()
+
+  span.SetAttributes(
+    attribute.Int("size", len(dah.RowsRoots)),
+    attribute.String("data_hash", hex.EncodeToString(dah.Hash())),
+  )
+  ...
+}
+```
+
+The child span in `ipld.Retriever.Reconstruct`:
+```go
+	ctx, span := tracer.Start(ctx, "reconstruct-square")
```

Comment:
According to [these docs](https://opentelemetry.io/docs/reference/specification/trace/semantic_conventions/#event-name-reuse-prohibition), traces must have unique names:

> A new event MUST NOT be added with the same name as an event that existed in the past but was renamed (with a corresponding schema file).

> When introducing a new event name check all existing schema files to make sure the name does not appear as a key of any “rename_events” section (keys denote old event names in rename operations).

Should we be concerned that it may be difficult to maintain uniqueness of trace (and metric) names in the long-run if they are scattered throughout our codebase? Do you think it would aid maintainability to have a single file of constant trace (and metric) names so there is one location to check for existing names prior to adding a new name? Or would a centralized trace (and metric) name file be cumbersome to use?

Reply:
Good point. Note that each span name is prepended with a namespace of created root tracer, meaning that the full name of this span would be `ipld.reconstruct-square`

Reply:
Added more info about this.